### PR TITLE
1.6

### PR DESCRIPTION
**Pull request by Chris-Castle**
Bug Fix on delete and rules based redirect

### CODE REVIEW

**Comment by thecadams** [comment]
Chris,
This is an amazing amount of work - thanks so much for contributing! Unfortunately I don't have a windows environment right now, so I'm wondering - can you point me in the right direction to test it? Would you recommend using an Azure VM and putting Sitecore on there, or something else?

File: README.md [hunk 1]
```diff
@@ -6,6 +6,17 @@ An improved version of the Sitecore 301 redirect module. Original version was cr
 Improvements by [Chris Adams](http://www.cadams.com.au/), [Max Slabyak](https://github.com/maxslabyak) and [Mark Wiseman](https://github.com/mawiseman)
```
Comment (thecadams) [inline README.md]: Should add your name here!

File: data/packages/301RedirectModule-1.5.xml [hunk 2]
```diff
@@ -16,6 +16,19 @@ An improved version of the Sitecore 301 redirect module. Original version was cr
 Improvements by [Chris Adams](http://www.cadams.com.au/), [Max Slabyak](https://github.com/maxslabyak) and [Mark Wiseman](https://github.com/mawiseman)
```
Comment (thecadams) [inline data/packages/301RedirectModule-1.5.xml]: And your name here!

File: data/packages/301RedirectModule-1.6.xml [hunk 3]
```diff
@@ -1,202 +1,303 @@
...
+      <Revision />
+      <License />
+      <Comment>A modification of a package contributed to Sitecore Shared Source by Chris Castle. Original is available at http://trac.sitecore.net/301RedirectModule</Comment>
+      <Attributes />
+      <Readme>301RedirectModule
+=================
+
+An improved version of the Sitecore 301 redirect module. Original version was created by Chris Castle, available at [http://trac.sitecore.net/301RedirectModule](http://trac.sitecore.net/301RedirectModule)
+
+Improvements by [Chris Adams](http://www.cadams.com.au/), [Max Slabyak](https://github.com/maxslabyak) and [Mark Wiseman](https://github.com/mawiseman)
```
Comment (thecadams) [inline data/packages/301RedirectModule-1.6.xml]: And here

File: source/Handlers/AutoCreateRedirectOnMove.cs [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+            {    
+                string oldPath = LinkManager.GetItemUrl(oldParent).Replace("/sitecore/shell", "") + "/" + LinkManager.GetItemUrl(item).Split('/').Last();
+                Database db = Sitecore.Configuration.Factory.GetDatabase("master");
+                // Get the generated folder underneath the redirects folder.  It is a bucketed item.
+                Item parentItem = db.GetItem(new ID("{46CE2092-FF8D-454E-B826-A2ADDB7E0BA3}"));
+
+                if (parentItem != null)
+                {
+                    //Now we need to get the template from which the item is created (Redirect Url)
+                    TemplateItem template = db.GetTemplate(new ID("{B5967A68-7F70-42D3-9874-0E4D001DBC20}"));
```
Comment (thecadams) [inline source/Handlers/AutoCreateRedirectOnMove.cs]: Probably worth extracting this to a constant

File: source/Handlers/AutoCreateRedirectOnMove.cs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+                    {
+                        // Create the item
+                        Item newItem = parentItem.Add(ItemUtil.ProposeValidItemName(oldPath.Replace('/',' ')), template);
+
+                        newItem.Editing.BeginEdit();
+                        try
+                        {
+                            // Assign values to the fields of the new item
+                            newItem.Fields["Requested Url"].Value = oldPath;
+                            newItem.Fields["Response Status Code"].Value = "{3184B308-C050-4A16-9F82-D77190A28F0F}";  // 301
```
Comment (thecadams) [inline source/Handlers/AutoCreateRedirectOnMove.cs]: And this one

File: source/Handlers/AutoCreateRedirectOnMove.cs [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+        private void CreateRedirectItem(Item item, Item oldParent)
+        {
+            // we only want a redirect on pages and media assets
+            if (oldParent.Paths.IsContentItem || oldParent.Paths.IsMediaItem)
+            {    
+                string oldPath = LinkManager.GetItemUrl(oldParent).Replace("/sitecore/shell", "") + "/" + LinkManager.GetItemUrl(item).Split('/').Last();
+                Database db = Sitecore.Configuration.Factory.GetDatabase("master");
+                // Get the generated folder underneath the redirects folder.  It is a bucketed item.
+                Item parentItem = db.GetItem(new ID("{46CE2092-FF8D-454E-B826-A2ADDB7E0BA3}"));
```
Comment (thecadams) [inline source/Handlers/AutoCreateRedirectOnMove.cs]: Probably worth extracting this to a constant

**Comment by Chris-Castle** [comment]
Hi Chris,

I don’t expect much impact because of the way the code works.  The module itself does nothing unless you ask for a page that Sitecore can’t resolve.  Then, it looks for a direct match, pattern match or rules match in that order.  So if you are primarily using direct matches, the new code rarely runs.  The only think I considered doing is using the index to look for a match instead of the Sitecore APIs against the content tree.

As for the other new code such as the event handler on item move, these are simple and have very little performance impact.

Thanks,
Chris

Chris Castle ● Sitecore  ● o:281-516-9861 ● m:415-385-1830 ● chc@sitecore.net<mailto:chc@sitecore.net>

**Comment by thecadams** [comment]
Looks good then, merging. Congrats and thanks again! :beer:

**Pull request merged**